order_book/src/order.rs

```diff
@@ -0,0 +1,289 @@
+                // the entire amount.
+                // The assert here makes sense since we should only get here if base_tokens is a
+                // range containing only self.sci.pseudo_output_amount.value
+                assert!(base_tokens == self.sci.pseudo_output_amount.value);
+
+                // In a non-partial swap, the fulfiller need to provide the amount specified in
+                // the (only, for now) required output.
+                Ok(self.sci.required_output_amounts[0].value)
+            }
+
+            (num_required_outputs @ (0 | 1), 1) => {
+                // Single partial output or a single partial output + change amount.
+                // The fact that the required output is treated as change has been verified when
+                // the Order was created.
+
+                // The amount we are taking must be above the minimum fill value. It is expected
+                // to be, since we checked base_range at the beginning.
+                assert!(base_tokens >= input_rules.min_partial_fill_value);
+
+                // The amount we are taking must be below the maximum available. It is expected
+                // to be, since we checked base_range at the beginning.
+                let mut max_available_amount = self.sci.pseudo_output_amount.value;
+                if num_required_outputs == 1 {
+                    assert!(max_available_amount > self.sci.required_output_amounts[0].value);
+                    max_available_amount -= self.sci.required_output_amounts[0].value;
+                };
+                assert!(base_tokens <= max_available_amount);
+
+                // The ratio being filled
+                let fill_fraction_num: u128 = base_tokens as u128;
+                let fill_fractions_denom = self.sci.pseudo_output_amount.value as u128;
+
+                // Calculate the number of counter tokens we need to return as change to the
+                // offerer of the SCI. It is calculated as a fraction of the partial fill
+                // output.
+                let (amount, _) = input_rules.partial_fill_outputs[0].reveal_amount()?;
+                let num_128 = amount.value as u128 * fill_fraction_num;
+                // Divide and round down
+                Ok((num_128 / fill_fractions_denom) as u64)
+            }
+
+            _ => Err(Error::UnsupportedSci(format!(
+                "Unsupported number of required/partial outputs {}/{}",
+                input_rules.required_outputs.len(),
+                input_rules.partial_fill_outputs.len()
+            ))),
+        }
+    }
+}
+
+impl TryFrom<SignedContingentInput> for Order {
+    type Error = Error;
+
+    fn try_from(sci: SignedContingentInput) -> Result<Self, Self::Error> {
+        let timestamp = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .map_err(|_| Error::Time)?
+            .as_nanos()
+            .try_into()
+            .map_err(|_| Error::Time)?;
```

Comment:
This should probably be a pair of `(SignedContingentInput, Instant)` to be testable.